File: src/common/protocol.ts
import fs from 'node:fs';

export function readBase64File(path: string): Buffer {
  const text = fs.readFileSync(path, { encoding: 'utf-8' });
  const compact = text.split(/\s+/).join('');
  const padding = (4 - (compact.length % 4)) % 4;
  const padded = padding ? compact + '='.repeat(padding) : compact;
  return Buffer.from(padded, 'base64');
}

export function parseRequest(raw: Buffer): [Buffer, Buffer] {
  if (raw.length < 4) throw new Error('File too short: missing 4-byte length prefix for blob1');
  const headerLen = raw.readUInt32LE(0);
  if (raw.length < 4 + headerLen)
    throw new Error('blob1 length in header is inconsistent with actual size');
  const blob1 = raw.subarray(4, 4 + headerLen);
  const blob2 = raw.subarray(4 + headerLen);
  if (blob2.length < 32)
    throw new Error('blob2 too short: missing 32-byte AES key appended at the end');
  return [blob1, blob2];
}

export function parseHeaderBlob1(blob1: Buffer): {
  prefix: Buffer;
  session_id: Buffer;
  udid_raw: Buffer;
  response_key: Buffer;
  auth_key: Buffer;
} {
  if (blob1.length < 112)
    throw new Error('blob1 too short to contain the required 112 fixed bytes');
  return {
    prefix: blob1.subarray(0, blob1.length - 112),
    session_id: blob1.subarray(blob1.length - 112, blob1.length - 96),
    udid_raw: blob1.subarray(blob1.length - 96, blob1.length - 80),
    response_key: blob1.subarray(blob1.length - 80, blob1.length - 48),
    auth_key: blob1.subarray(blob1.length - 48),
  };
}

export function udidRawToCanonicalString(udidRaw: Buffer): string {
  const hx = udidRaw.toString('hex');
  return `${hx.slice(0, 8)}-${hx.slice(8, 12)}-${hx.slice(12, 16)}-${hx.slice(
    16,
    20,
  )}-${hx.slice(20, 32)}`;
}

export function deriveIvFromUdidString(udidString: string): Buffer {
  const udidNoHyphens = udidString.replace(/-/g, '');
  const ivAscii16 = udidNoHyphens.slice(0, 16);
  if (ivAscii16.length !== 16) throw new Error('Unexpected UDID: could not derive 16-byte IV');
  return Buffer.from(ivAscii16, 'utf-8');
}

File: common/protocol.py
import base64


def read_base64_file(path: str) -> bytes:
    with open(path, "r", encoding="utf-8", errors="ignore") as f:
        text = f.read()
    compact = "".join(text.split())
    padding = (-len(compact)) % 4
    if padding:
        compact += "=" * padding
    return base64.b64decode(compact)


def parse_request(raw: bytes) -> tuple[bytes, bytes]:
    if len(raw) < 4:
        raise ValueError("File too short: missing 4-byte length prefix for blob1")
    header_len = int.from_bytes(raw[:4], "little")
    if len(raw) < 4 + header_len:
        raise ValueError("blob1 length in header is inconsistent with actual size")
    blob1 = raw[4 : 4 + header_len]
    blob2 = raw[4 + header_len :]
    if len(blob2) < 32:
        raise ValueError("blob2 too short: missing 32-byte AES key appended at the end")
    return blob1, blob2


def parse_header_blob1(blob1: bytes) -> dict:
    if len(blob1) < 112:
        raise ValueError("blob1 too short to contain the required 112 fixed bytes")
    return {
        "prefix": blob1[:-112],
        "session_id": blob1[-112:-96],
        "udid_raw": blob1[-96:-80],
        "response_key": blob1[-80:-48],
        "auth_key": blob1[-48:],
    }


def udid_raw_to_canonical_string(udid_raw: bytes) -> str:
    hx = udid_raw.hex()
    return f"{hx[0:8]}-{hx[8:12]}-{hx[12:16]}-{hx[16:20]}-{hx[20:32]}"


def derive_iv_from_udid_string(udid_string: str) -> bytes:
    udid_no_hyphens = udid_string.replace("-", "")
    iv_ascii_16 = udid_no_hyphens[:16]
    if len(iv_ascii_16) != 16:
        raise ValueError("Unexpected UDID: could not derive 16-byte IV")
    return iv_ascii_16.encode("utf-8")
